eox_tenant/api/v1/permissions.py

```diff
@@ -1,13 +1,88 @@
+    used to give access to the eox-tenant API.
+    """
+    if settings.EOX_TENANT_LOAD_PERMISSIONS:
+        try:
+            content_type = ContentType.objects.get_for_model(User)
+            Permission.objects.get_or_create(  # pylint: disable=unused-variable
+                codename='can_call_eox_tenant',
+                name='Can access eox-tenant API',
+                content_type=content_type,
+            )
+        except ProgrammingError:
+            # This code runs when the app is loaded, if a migration has not been
+            # done a ProgrammingError exception is raised.
+            # we are bypassing those cases to let migrations run smoothly.
+            pass
 
 
 class EoxTenantAPIPermission(permissions.BasePermission):
-    """Only allows super user."""
+    """
+    Defines a custom permissions to access eox-tenant API.
+    These permissions make sure that a token is created with the client credentials of the same site is
+    being used on, and the user has a valid SignUp source for the site.
+    """
 
     def has_permission(self, request, view):
-        if request.user.is_superuser:
+        """
+        To grant access, checks if the requesting user:
+            1) it's a staff user
+            2) it's allowed to perform operations on the site (check user signUp sources)
+            3) it's calling the API from a site authorized by the auth application or client
+            4) has can call eox-tenant API permission
+        """
+        host = request.get_host()
+        user = request.user
+        user_sources = user.usersignupsource_set.all()
+
+        if user.is_staff:
             return True
+
+        if not self.is_authorized_source_for_user(host, user_sources):
+            return False
+
+        try:
+            application_uri_allowed = request.auth.application.redirect_uri_allowed(request.build_absolute_uri('/'))
+        except Exception:  # pylint: disable=broad-except
+            application_uri_allowed = False
+
+        try:
+            client_url_allowed = host in request.auth.client.url
+        except Exception:  # pylint: disable=broad-except
+            client_url_allowed = False
+
+        if client_url_allowed or application_uri_allowed:
+            return user.has_perm('auth.can_call_eox_tenant')
+
+        # If we get here either someone is using a token created on one site in a different site
+        # or there was a missconfiguration of the oauth client.
+        # we return the most basic message to prevent leaking important information .
+        raise exceptions.NotAuthenticated(detail="Invalid token")
+
+    @staticmethod
+    def is_authorized_source_for_user(request_host, user_signup_sources):
+        """
+        Returns True if the current user has a valid signUp source to make requests
+        to the current site.
+        """
+        authorized_sources = getattr(settings, 'EDNX_ACCOUNT_REGISTRATION_SOURCES', [request_host])
```

Comment:
If I understand this correctly, when there is no EDNX_ACCOUNT_REGISTRATION_SOURCES at the site, then authorized_sources is the host and thus this check will always pass. Right?

Reply:
This would be OK for me. But I'm also wondering, should we for this to have a registered source in the site and that is it. We don't need to support the feature EDNX_ACCOUNT_REGISTRATION_SOURCES in here.

The cost of not doing it is that every single site needs its own user to call this API. It does not sound so bad.

What I think is worse is that sites that do not use the `user_signup_sources` feature are completely blocked from this.
Now I get the sense that we should make validating the `user_signup_sources` optional according to a setting and leave the ednx_account... completely out of it.

What are your thoughts on this @Alec4r ?

Reply:
@felipemontoya I thought that all the sites were going to use the `user_signup_sources` feature, but if that's not the case, then I think we should remove the signup source validation. Do you think it's really worth it to leave it as an option?

In fact, the eox-core permission protocol doesn't make this signup source validation, it's something extra that I added because I thought it made sense https://github.com/eduNEXT/eox-core/blob/master/eox_core/api/v1/permissions.py#L32

Reply:
This makes sense but only in the case of a large multi-tenant installation such as our saas service. I would not want to make everyone have to fight this restriction for something that is pure edunext service.

Reply:
@felipemontoya I removed the SignUp source verification